File: programmers/kakao/level3/길_찾기_게임.py
import sys
sys.setrecursionlimit(10**6)

class Node:
    def __init__(self, id, x, y):
        self.id = id
        self.x = x
        self.y = y
        self.left = None
        self.right = None

    def __lt__(self, other):
        if (self.y == other.y):
            return self.x < other.x
        return self.y > other.y

def addNode(parent, child):
    if child.x < parent.x:
        if parent.left is None:
            parent.left = child
        else:
            addNode(parent.left, child)
    else:
        if parent.right is None:
            parent.right = child
        else:
            addNode(parent.right, child)

def preorder(ans, node):
    if node == None: return
    ans.append(node.id)
    preorder(ans, node.left)
    preorder(ans, node.right)

def postorder(ans, node):
    if node == None: return
    postorder(ans, node.left)
    postorder(ans, node.right)
    ans.append(node.id)

def solution(nodeinfo):
    size = len(nodeinfo)
    node_lst = []
    for i in range(size):
        node_lst.append(Node(i+1, nodeinfo[i][0], nodeinfo[i][1]))

    node_lst.sort()
    root = node_lst[0]
    for i in range(1, size):
        addNode(root, node_lst[i])

    answer = [[], []]
    preorder(answer[0], root)
    postorder(answer[1], root)
    return answer


nodeinfo = [[5,3],[11,5],[13,3],[3,5],[6,1],[1,3],[8,6],[7,2],[2,2]]
print(solution(nodeinfo)) # [[7,4,6,9,1,8,5,2,3],[9,6,5,8,1,4,3,2,7]]

File: Java/programmers/kakao/level3/길찾기게임.java
package CodeTest.Java.programmers.kakao.level3;

import java.util.*;

public class 길찾기게임 {
    static int[][] answer;
    static int idx;

    public int[][] solution(int[][] nodeinfo) {
        int n = nodeinfo.length;
        
        Node[] node = new Node[n];
        for (int i=0; i<n; i++) {
            node[i] = new Node(nodeinfo[i][0], nodeinfo[i][1], i+1, null, null);
        }

        Arrays.sort(node, new Comparator<Node>() {
            @Override
            public int compare(Node n1, Node n2) {
                if(n1.y == n2.y) return n1.x - n2.x;
                else return n2.y - n1.y;
            }
        });

        Node root = node[0];
        for (int i=1; i<n; i++) {
            insertNode(root, node[i]);
        }

        answer = new int[2][n];
        idx = 0;
        preorder(root);
        idx = 0;
        postorder(root);
        return answer;
    }

    private static void insertNode(Node parent, Node child) {
        if (parent.x > child.x) {
            if (parent.left == null) parent.left = child;
            else insertNode(parent.left, child);
        } else {
            if (parent.right == null) parent.right = child;
            else insertNode(parent.right, child);
        }
    }

    private static void preorder(Node root) {
        if (root != null) {
            answer[0][idx++] = root.value;
            preorder(root.left);
            preorder(root.right);
        }
    }

    private static void postorder(Node root) {
        if (root != null) {
            postorder(root.left);
            postorder(root.right);
            answer[1][idx++] = root.value;
        }
    }

    private class Node {
        int x, y;
        int value;
        Node left, right;

        private Node(int x, int y, int value, Node left, Node right) {
            this.x = x;
            this.y = y;
            this.value = value;
            this.left = left;
            this.right = right;
        }
    }
}
